app/src/org/commcare/adapters/MenuAdapter.java

```diff
@@ -213,25 +210,37 @@ public void setupImageView(ImageView mIconView, MenuDisplayable menuDisplayable,
         }
     }
 
-    protected void setupBadgeView(View menuListItem, MenuDisplayable menuDisplayable) {
-        View badgeView = menuListItem.findViewById(R.id.badge_view);
-        compositeDisposable.add(
-                menuDisplayable.getTextForBadge(asw.getEvaluationContext(menuDisplayable.getCommandID()))
-                        .subscribeOn(Schedulers.computation())
-                        .observeOn(AndroidSchedulers.mainThread())
-                        .subscribe(badgeText -> updateBadgeView(badgeView, badgeText),
-                                throwable -> UserfacingErrorHandling.createErrorDialog(context, throwable.getLocalizedMessage(), true)
-                        )
-        );
+    private void setupBadgeView(View badgeView, MenuDisplayable menuDisplayable, int position) {
+        badgeView.setVisibility(View.GONE);
+        badgeView.setTag(position);
+
+        if (badgeCache.get(position) != null) {
+            updateBadgeView(badgeView, badgeCache.get(position));
+        } else {
+            compositeDisposable.add(
+                    menuDisplayable.getTextForBadge(asw.getEvaluationContext(menuDisplayable.getCommandID()))
+                            .subscribeOn(Schedulers.computation())
+                            .observeOn(AndroidSchedulers.mainThread())
+                            .subscribe(badgeText -> {
+                                        // Make sure that badgeView corresponds to the right position and update it
+                                        if (((int)badgeView.getTag()) == position) {
```

Comment:
@shubham1g5 what makes this check necessary?

Reply:
If we don't do this and scroll through listview, we can get the result of the async request after that particular view has been recycled for another position. For ex. I made a request for badgeView which was at position 5 and immediately scroll downwards which will cause that particular badgeview to be recycled and get used for some other position say 15. So when we get back the result from the async request for data at position 5, we will update the badgeview for position 15 with position 5 data until we finally get the result for position 15 data.

Reply:
Ah cool makes perfect sense